enhancements/single-node/single-node-openshift-with-workers.md

```diff
@@ -0,0 +1,445 @@
+running conformance tests post-reboot.
+
+- Add tests on both cloud / "none"-platform that check that a single-node cluster
+with additional workers recovers after an upgrade by running conformance tests
+post-upgrade.
+
+TODO: Describe day-1-workers tests?
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+TBD
+
+#### Dev Preview -> Tech Preview
+
+TBD
+
+#### Tech Preview -> GA
+
+TBD
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+In the non-goals section it's mentioned that this enhancement does not apply to
+clusters which have been installed prior to the enhancement, so their upgrade
+is not discussed.
+
+This enhancement, to the best of my knowledge, should have no problems
+persisting across cluster upgrades. The Test Plan section describes how this
+will be tested.
+
+### Version Skew Strategy
+
+Does not apply, to the best of my understanding.
+
+### Operational Aspects of API Extensions
+
+- The `ControlPlane` value of the `IngressPlacement` parameter value may not be
+used when the cluster's `ControlPlaneTopology` parameter is set to `External`,
+this combination is currently considered invalid. This may change in the future.
+
+- This API change only affects the defaulting behavior of the `IngressController` CR,
+it does not add any new capabilities to OCP, or give any more flexibility than there
+already was.
+
+- For administrators and support engineers, the `IngressController` is still the
+source of truth and where you need to look if you seek to understand the router
+placement in practice. Nothing has changed in that regard.
+
+#### Failure Modes
+
+- The `ControlPlane` value of the  `IngressPlacement` parameter value may not be
+used if the cluster's `ControlPlaneTopology` parameter is set to `External`.
+The Ingress Operator would treat such combination as if the `IngressPlacement`
+value is actually `Workers` instead, and log a warning to indicate that this
+invalid combination has been detected. The Ingress Operator should not fail
+to reconcile `IngressController` CRs due to this invalid combination.
+
+#### Support Procedures
+
+- For administrators and support engineers, the `IngressController` is still the source
+of truth and where you need to look for understanding the router placement in practice.
+Nothing has changed in that regard.
+
+## Implementation History
+
+Not yet applicable
+
+## Drawbacks
+
+- The pinning of the `IngressController` to the "master" pool is another change
+which would make single-node clusters slightly different from multi-node
+clusters, and any such difference is naturally not ideal.
+
+- The proposed defaulting behavior for the discussed `IngressController`
+parameters is complicated and dependent on three different parameters (infra
+topology, control-plane topology, and ingress placement) - such complexity
+would probably have to be documented in the CRD definitions and may confuse
+users.
```

Comment:
It will definitely need to be documented within the CRD definitions. But, this begs the question, if it's so complex, why are we doing it? What is the advantage here?

Is it better to just have a user configure the resources correctly with no defaulting? (I don't think we can actually do that because this API was already released, but it's good to think about what the ideal world would look like and note that)

The complexity here and the apparent magic that will happen behind the scenes, from a UX perspective, worries me. This would need to be very thoroughly documented to be understood, and even then, I think the behaviour could be surprising to users

Reply:
This is useful context, I didn't see it in the enhancement already when I read through, can we make sure to add it

Reply:
This can be resolved, thank you for the updates